### PR TITLE
Fix inconsistency in `ls` sort-order

### LINKED ISSUE #13267: `ls <dir>` no longer sorted by date
### Describe the bug

This appears to have changed/broken in 0.94, but (once again), I left my long-running script running for too long under 0.93.  When I skipped forward to 0.95, this broke my script, since it was trying to get the `ls <dir> | last` file from the directory.  That filename was no longer the same as it had been in 0.93.

### How to reproduce

```nu
> cd /
> ls
# sorted by filename
> ls (pwd)
# not sure what the sort-order is
```

### Expected behavior

Sort `ls` output by filename, even when a directory path is specified.

### Screenshots

_No response_

### Configuration

| key                | value                                    |
| ------------------ | ---------------------------------------- |
| version            | 0.95.0                                   |
| major              | 0                                        |
| minor              | 95                                       |
| patch              | 0                                        |
| branch             |                                          |
| commit_hash        |                                          |
| build_os           | linux-x86_64                             |
| build_target       | x86_64-unknown-linux-gnu                 |
| rust_version       | rustc 1.77.2 (25ef9e3d8 2024-04-09)      |
| rust_channel       | 1.77.2-x86_64-unknown-linux-gnu          |
| cargo_version      | cargo 1.77.2 (e52e36006 2024-03-26)      |
| build_time         | 2024-06-28 06:28:34 -04:00               |
| build_rust_channel | release                                  |
| allocator          | mimalloc                                 |
| features           | default, sqlite, system-clipboard, trash |
| installed_plugins  |                                          |

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by userwiths**
Fixes #13267 

As we can see from the bisect done in the comments.
Bisected to https://github.com/nushell/nushell/pull/12625 / https://github.com/nushell/nushell/commit/460a1c8f874b503cb4140619838be483600f00ce

We can see that this update brought the use of `read_dir` and for it, it is mentioned in the [rust docs](https://doc.rust-lang.org/std/fs/fn.read_dir.html#platform-specific-behavior) that it does **not** provide any specific order of files.
As was the advice there, I went and applied a manual `sort` to the entries and tested it manually on my local machine.

If required I could probably try and add tests for the order consistency, would need some time to find my way around them, so I'm sending the PR first.

### CODE REVIEW

**Comment by fdncred** [comment]
This collects and therefore stops `ls` from streaming.

**Comment by userwiths** [comment]
Now the branch is using `itertools`'s `sorted` function. This required the implementation of a few methods for `ShellError` used for the comparison done by it.
Also added a test following the examples of the previous ones.

**Comment by sholderbach** [comment]
Using `itertools::sorted` doesn't change the fact that it has to consume the whole iterator first before it can emit the elements in the sorted order. The second question is what is the sorting comparator in use. Is it pure string sorting or follows some hierarchy?

If the platforms have some stronger guarantees about what their directory walking algorithms do, we could maybe avoid having to hold all entries back and instead could start yielding early?

**Comment by userwiths** [comment]
While I agree with you that currently we are still consuming the iterator, I fail to see any possible way to order/sort elements without knowing about them all first. And as `read_dir` currently does not guarantee the same order as `ls` (lexicographic I believe, should answer the first question) we have to sort them if we wish consistent order.

As for the "platforms" point, the fact that the docs mention possible change of behavior is a bit discouraging.

I'm open to suggestions, but the topics of streams/async/iterators are things in which I could hardly contribute without  (a lot of) guidance.

**Comment by sholderbach** [comment]
Mhh yeah, if the behavior pre #12625 also effictively collected the particular directory, then sorting one directory is probably fine.

**Comment by WindSoilder** [comment]
> Mhh yeah, if the behavior pre https://github.com/nushell/nushell/pull/12625 also effictively collected the particular directory, then sorting one directory is probably fine.

Ah, @sholderbach you make a good point!  I think `glob_from` will somehow invoke collect, and then sort result by filename

https://github.com/nushell/nushell/blob/497954d84c65e9c5a02cc40ef10ada5c060684c8/crates/nu-glob/src/lib.rs#L939-L967

**Comment by fdncred** [comment]
seems like currently there are two possibilities, one that's using `--threads` and one that doesn't. maybe we shouldn't collect on the one that uses threads?

**Comment by sholderbach** [comment]
Trying to sort the `ShellError` with no direct semantic meaning smells to me. Can we avoid having to add those impls by going back to sorting without itertools in the collect case. 

@fdncred is the codepath actually read lazily in the threading mode?

**Comment by fdncred** [comment]
@sholderbach I don't think so. 

I'm not convinced that `ls` streams at all. First of all, every mode is threading mode now. It's just either 1 thread or number_of_logical_processors threads. It definitely doesn't seem to return a list stream or a byte stream. Describe says it's a stream but I'm not convinced of that because I can't get it to output data in chunks. I'm guessing that filter_map is collecting it, but I could be wrong, it could be step 2 that isn't threaded.

This is a rough breakdown of how it works.
1. setup a mpsc channel
2. setup paths_peek, which is either the output of read_dir or glob_from. i don't believe either is multi-threaded.
3. create a threadpool of 1 or logical proc number
4. whatever is in paths_peek we parallelize that iterator with par_bridge(), which uses the threadpool we created, and send down filter_map
5. at the end of each filter_map run, we tx.send, rinse and repeat
6. the rx returns with into_pipeline_data

so, you would think that if there are 1000 items in paths_peek, par_bridge() would parallize them, and you can see that when you `ls -t dir_with_lots_of_files`. However, if you do `ls -t c:\windows\**` it'll take forever to return, if you have enough patience. so, the work in step 2 above is a limiting factor and should be improved on.

File: crates/nu-command/src/filesystem/ls.rs [hunk 1]
```diff
@@ -979,10 +980,16 @@ mod windows_helper {
...
     let iter = f.read_dir()?.map(|d| {
         d.map(|r| r.path())
             .map_err(|e| ShellError::IOError { msg: e.to_string() })
     });
-    Ok(Box::new(iter))
+
+    if use_threads {
+        Ok(Box::new(iter))
+    } else {
+        Ok(Box::new(sorted(iter)))
```
Comment (WindSoilder) [inline crates/nu-command/src/filesystem/ls.rs:993]: Let's collect the iterator and sort it, so we don't need to use `itertools::sorted`.  And no changes is required on `ShellError`.

**Review by WindSoilder — CHANGES_REQUESTED** [review]
Thank you for the changes!  I'd like to push forward it.
Just left a comment

File: crates/nu-command/src/filesystem/ls.rs (hunk 1 above)
Comment (userwiths) [reply crates/nu-command/src/filesystem/ls.rs]: Ok, I think now it should be as you describe it.

File: crates/nu-protocol/src/errors/shell_error.rs [hunk 2]
```diff
@@ -1459,6 +1459,7 @@ On Windows, this would be %USERPROFILE%\AppData\Roaming"#
     },
 }
 
+// TODO: Implement as From trait
```
Comment (WindSoilder) [inline crates/nu-protocol/src/errors/shell_error.rs:1462]: I don't think we need this comment

File: crates/nu-command/src/filesystem/ls.rs [hunk 3]
```diff
@@ -980,9 +980,18 @@ mod windows_helper {
 fn read_dir(
     f: &Path,
```
Comment (WindSoilder) [inline crates/nu-command/src/filesystem/ls.rs:981]: Still, I think `use_thread` parameter is required.  We don't want to sort items if it's using thread.

**Review by WindSoilder — CHANGES_REQUESTED** [review]
Thanks!  I left some comments

File: crates/nu-command/src/filesystem/ls.rs [hunk 4]
```diff
@@ -979,10 +979,22 @@ mod windows_helper {
...
-            .map_err(|e| ShellError::IOError { msg: e.to_string() })
-    });
-    Ok(Box::new(iter))
+    let mut items = f
+        .read_dir()?
+        .map(|d| {
+            d.map(|r| r.path())
+                .map_err(|e| ShellError::IOError { msg: e.to_string() })
+        })
+        .collect::<Vec<_>>();
```
Comment (WindSoilder) [inline crates/nu-command/src/filesystem/ls.rs:990]: Should collect only if `!use_threads`.

**Review by WindSoilder — APPROVED** [review]
Thank you for going through this!  Looks good to me!

**Pull request merged**